apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+# Refactoring code.org javascript
+
+The code base has been around for a while and a lot has changed. While there is
+still a lot of spaghetti, it's at least now in a place where we have the
+opportunity to refactor the code without changing build systems or doing any other
+super complicated thing.
+
+
+## The case for refactoring ##
+
+code.org has multiple websites, each of which have their own set of features
+that rely on rich client side javascript. Most of the javascript for these
+different sites has ended up in the `apps/src` directory, whether or not they
+relate to or depend on each other. For historical reasons, this code is not
+organized well, which causes a number of problems for us:
+
+- it is difficult to understand the execution flow across module bounderies
+  without resorting to step debugging
+- it is difficult to understand how modules could be efficiently broken up into
+  different bundles in a way that reduces page load times
+- it is difficult for newcomers to understand how the code base is organized
+  (it's not!)
+- it is difficult to figure out how/where to add code in a way that doesn't add
+  to the spaghetti that is already there
+
+The goal of refactoring would be to make all of the above much less difficult
+and time consuming.
+
+## Key challenges to refactoring ##
+
+Our codebase has morphed over time to use a number of different
+strategies/libraries/frameworks for managing state, business logic, and ui. If
+we were using a single strategy/library/framework, it would be easier to come up
+with a filesystem and module architecture that followed the best practices of
+that framework.
+
+At the moment, we have collectively decided to invest heavily in adopting react
+for the UI, and redux for the application state. Unfortunately, given the size
+of our codebase, it's unrealistic to deprecate or rewrite all the code that was
+written using jquery, ejs templates, and raw javascript. On top of that, we use
+a number of third party libraries (piskel, blockly, etc.) for large parts of our
+product, and these libraries don't necessarily use react or redux.
+
+We also have a lot of code that doesn't really involve UI or state, but which
+provides core functionality for our product, such as the JS interpreter.
```

Comment:
The JS Interpreter is [just another third-party library](https://github.com/NeilFraser/JS-Interpreter) and should be treated as such.  Or are you referring to our wrapper code around it?  We're going to have some sort of adapter for many of our dependencies.